### PR TITLE
`Adaptive learning`: Unlink edited competencies from standardised competencies

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Competencies and prerequisites can be created using standardised competencies. If these imported competencies or prerequisites are edited later on, they should loose their link to the standardised competency, as they do no longer resemble the standardised competency.

### Description

When a competency is edited that is linked to a standardised competency, a warning is shown above the submit button that this change will unlink the competency. And on submit the competency is then actually unlinked.

### Steps for Testing

Prerequisites:
- 1 Instructor
- Standardised competencies (e.g. on ts3)

1. Log in to Artemis
2. Import a competency and/or prerequisite based on a standardised competency
3. Edit the imported competency/prerequisite
4. Edit it again and check that there is no longer a warning as it is no longer linked

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

### Screenshots

Bildschirmfoto 2024-10-01 um 14 03 05

## Release Notes

- **New Features**
	- Added a warning message for users when editing competencies, indicating that it will unlink standardized competencies.
	- Introduced a new property to the competency form to indicate linked standardized competencies.
	- Enhanced user feedback with a conditional alert for linked standardized competencies during form submission.

- **Bug Fixes**
	- Enhanced the handling of edit mode checks in various components for improved reliability.

- **Documentation**
	- Updated localization files for both German and English to reflect new warning messages and additional fields in the competency management sections.

### CODE REVIEW

File: src/main/webapp/i18n/de/competency.json [hunk 1]
```diff
@@ -209,6 +209,9 @@
                     "noOfParticipants": "Anzahl der Teilnehmenden",
                     "participationRate": "Teilnahmerate"
                 }
+            },
+            "edit": {
+                "unlinkStandardizedCompetenciesWarning": "Das Editieren wird dazu führen, dass die Verbindung zu standardisierten Kompetenzen gelöst wird!"
```
Comment (az108) [inline src/main/webapp/i18n/de/competency.json:214]: I would write: "Das Editieren führt zum lösen der Verbindung zu standardisierten Kompetenzen!" or similar but currently its tiny bit weird.

**Review by az108 — CHANGES_REQUESTED** [review]
Overall Code 👍

File: src/main/webapp/i18n/de/competency.json (hunk 1 above)
Comment (pzdr7) [reply src/main/webapp/i18n/de/competency.json:214]: I'd prefer "Bearbeiten" over "Editieren"

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Please check the failing client tests. You can set the inputs using

```ts
fixture.componentRef.setInput('attributeName', value);
```

File: src/main/webapp/app/course/competencies/forms/course-competency-form.component.ts [hunk 2]
```diff
@@ -55,20 +55,14 @@ export interface CourseCompetencyFormData {
...
-    lecturesOfCourseWithLectureUnits: Lecture[] = [];
-    @Input()
-    averageStudentScore?: number;
-    @Input()
-    hasCancelButton: boolean;
+    isEditMode = input<boolean>(false);
+    isInConnectMode = input<boolean>(false);
+    isInSingleLectureMode = input<boolean>(false);
+    hasLinkedStandardizedCompetency = input<boolean>(false);
+    courseId = input<number>();
```
Comment (JohannesWt) [inline src/main/webapp/app/course/competencies/forms/course-competency-form.component.ts:62]: Use required
```suggestion
    courseId = input.required<number>();
```

**Review by JohannesWt — CHANGES_REQUESTED** [review]
Code looks good just a small remark

**Review by pzdr7 — APPROVED** [review]
Code

File: src/main/webapp/app/course/competencies/forms/course-competency-form.component.ts [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import { Component, EventEmitter, Input, Output } from '@angular/core';
+import { Component, EventEmitter, Output, input } from '@angular/core';
```
Comment (Jan-Thurner) [inline src/main/webapp/app/course/competencies/forms/course-competency-form.component.ts:1]: Is there a reason for changing the order here?

**Review by Jan-Thurner — APPROVED** [review]
The code looks good to me. I left one question in the comments.

File: src/main/webapp/app/course/competencies/forms/course-competency-form.component.ts (hunk 3 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/course/competencies/forms/course-competency-form.component.ts:1]: Yes, input is now no longer capitalised (signal vs. `@Input()`) and our linter wants all capitalised words before the others

**Review by az108 — APPROVED** [review]
Tested locally and works as expected 👍

**Review by cremertim — APPROVED** [review]
Code lgtm

**Pull request closed without merging**